d2/src/main/java/com/linkedin/d2/balancer/util/CanaryDistributionProvider.java

```diff
@@ -0,0 +1,27 @@
+package com.linkedin.d2.balancer.util;
+
+import com.linkedin.d2.D2CanaryDistributionStrategy;
+
+
+/**
+ * Canary Distribution Provider is supposed to provide information about canary distributions.
+ */
+public interface CanaryDistributionProvider {
```

Comment:
And if we agreed to implement this in pegasus, i feel like it'd be sufficient to just have the class, rather than class + interface. The polymorphism is already presented in D2CanaryDistributionStrategy

Reply:
we discussed this with the team in one of the standups last year, and the team concluded with putting the implementation class in container for the reasons:
1. the distribution logic may change in the future to involve more LI-specific info. We don't want to wait till then to move it with backward-compatibility problems. The suggestion is to abstract the operation and not the data. This will help us use the same abstraction layer to implement different canary strategies depending on future data that is not anticipated today and not readily available in pegasus. If we have to do the canary picking logic in pegasus - we have to build the abstraction around how to pass the data into pegasus, and that abstraction is easily broken if we need additional data for future canary use case that is not covered today.
2. the canary distribution logic could be re-used by other features in both pegasus and container in future too.

The reason to have inter + class is because D2CanaryDistributionStrategy only defines different canary strategies (ramp, target host, target apps) but there could be different implementations in the distribution logic. My current implementation uses XXHash, while future implementations may add different hash logic or depend on completely different mechanisms.

Reply:
IMO, even if we may need LI-specific info in the future, we should still strive to provide at least a basic implementation in pegasus, and have a wrapper in container to extend the basic implementation. For example, `BaseDarkClusterDispatcherImpl` is the concrete class of `DarkClusterDispatcher`, which is extended by container to include LI-specific info like IC. Otherwise, this canary config feature is missing a core component in the open-source version.

Reply:
make sense. I will add a very basic implementation in pegasus and let the container one extends it.

Reply:
added. thanks.